cmake/modules/extensions.cmake

```diff
@@ -2524,59 +2558,41 @@ function(zephyr_get variable)
        ("${GET_VAR_SYSBUILD}" STREQUAL "GLOBAL" OR sysbuild_main_app)
     )
       get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
-      set(used_global true)
-    endif()
-  endif()
-
-  if(DEFINED sysbuild_${variable})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} ${sysbuild_${variable}})
-    else()
-      set(${variable} ${sysbuild_${variable}} PARENT_SCOPE)
-      return()
-    endif()
-  endif()
-  if(SYSBUILD AND GET_VAR_MERGE AND NOT used_global AND "${GET_VAR_SYSBUILD}" STREQUAL "GLOBAL")
-    get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
-    list(APPEND ${variable} ${sysbuild_${variable}})
-  endif()
-  if(DEFINED CACHE{${variable}})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} $CACHE{${variable}})
-    else()
-      set(${variable} $CACHE{${variable}} PARENT_SCOPE)
-      return()
-    endif()
-  endif()
-  if(DEFINED ENV{${variable}})
-    if(GET_VAR_MERGE)
-      list(APPEND ${variable} $ENV{${variable}}})
-    else()
-      set(${variable} $ENV{${variable}} PARENT_SCOPE)
-      # Set the environment variable in CMake cache, so that a build invocation
-      # triggering a CMake rerun doesn't rely on the environment variable still
-      # being available / have identical value.
-      set(${variable} $ENV{${variable}} CACHE INTERNAL "")
-    endif()
-
-    if(DEFINED ${variable} AND NOT "${${variable}}" STREQUAL "$ENV{${variable}}")
-      # Variable exists as a local scoped variable, defined in a CMakeLists.txt
-      # file, however it is also set in environment.
-      # This might be a surprise to the user, so warn about it.
-      message(WARNING "environment variable '${variable}' is hiding local "
-                      "variable of same name.\n"
-                      "Environment value (in use): $ENV{${variable}}\n"
-                      "Local scope value (hidden): ${${variable}}\n"
-      )
     endif()
+  else()
+    set(sysbuild_${variable})
+  endif()
+
+  set(scopes "sysbuild;CACHE;ENV;current")
+  foreach(scope IN LISTS scopes)
+    foreach(var ${GET_VAR_VAR})
+      zephyr_var_name("${var}" "${scope}" expansion_var)
+      if(DEFINED expansion_var)
+        string(CONFIGURE "${expansion_var}" scope_value)
```

Comment:
I'm curious about the use of CONFIGURE here. This needs to resolve `$CACHE` and `$ENV` references too, but I don't see any documentation in https://cmake.org/cmake/help/latest/command/configure_file.html that says that should work. Only `${VAR}` and `@VAR@` are explicitly supported. Is this relying on an implementation detail of cmake?

Reply:
That's a very good point.

So I opened this: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8500